altos-rust/altos-core/libs/heap/free_list_allocator/src/test.rs

```diff
@@ -0,0 +1,129 @@
+ */
+
+use std::vec::Vec;
+use free_list;
+
+// TestMemory is a helper type which represents a chunk of memory used in testing the allocator
+// It uses a vector to create the necessary memory and manages deallocation of the vector
+pub struct TestMemory {
+    heap: *mut u8,
+    vector_length: usize,
+    vector_capacity: usize,
+}
+
+impl TestMemory {
+    pub fn new(heap_size: usize) -> TestMemory {
+        let mut heap: Vec<u8> = vec![0; heap_size];
+        let heap_start = &mut heap[0] as *mut u8;
+        let heap_capacity = heap.capacity();
+        ::core::mem::forget(heap);
+        TestMemory {
+            heap: heap_start,
+            vector_length: heap_size,
+            vector_capacity: heap_capacity,
+        }
+    }
+
+    pub fn get_heap(&self) -> *mut u8 {
+        self.heap
+    }
+
+    fn get_memory_size(&self) -> usize {
+        self.vector_length
+    }
+}
+
+impl Drop for TestMemory {
+    fn drop(&mut self) {
+        unsafe {
+            // Reform the vector from component parts in order to make sure it's deallocated
+            let heap = Vec::from_raw_parts(self.heap, self.vector_length, self.vector_capacity);
+            drop(heap);
+        }
+    }
+}
+
+// TestFreeList is a helper type for free lists, which utilizes TestMemory
+// Provides some functions which help make it easier to check certain properties of the free list
+pub struct TestFreeList {
+    pub free_list: free_list::FreeList,
+    test_memory: TestMemory,
+}
+
+impl TestFreeList {
+    fn new(test_memory: TestMemory) -> TestFreeList {
+        let mut free_list = free_list::FreeList::new();
+        free_list.init(test_memory.get_heap() as usize, test_memory.get_memory_size());
+        TestFreeList {
+            free_list: free_list,
+            test_memory: test_memory,
+        }
+    }
+
+    pub fn get_heap_start(&self) -> *mut u8 {
+        self.test_memory.get_heap()
+    }
+
+    // Helper function which returns the sum of the block_sizes for every BlockHeader in the list
+    pub fn sum_free_block_memory(&self) -> usize {
+        let mut current = self.free_list.head;
+        let mut sum: usize = 0;
+        while !current.is_null() {
+            sum += unsafe { (*current).block_size };
+            current = unsafe { (*current).next_block };
+        }
+        sum
+    }
+
+    // Helper function to get the number of BlockHeader nodes in the list
+    pub fn count_free_blocks(&self) -> usize {
+        let mut current = self.free_list.head;
+        let mut num_blocks = 0;
+        while !current.is_null() {
+            num_blocks += 1;
+            current = unsafe { (*current).next_block };
+        }
+        num_blocks
+    }
+
+    // Helper function to check that every block in the list satisfies some condition
+    // Returns false if the condition returns false for any node
+    pub fn each_free_block_satisfies(&self, condition: &Fn(*mut free_list::BlockHeader) -> bool)
```

Comment:
For function arguments, the signature should use a generic over the `Fn` trait. So
```rust
fn each_free_block_satisfies<F: Fn(*mut BlockHeader) -> bool>(&self, condition: F) -> bool
```